### PR TITLE
add 'version' to the package.yaml file

### PR DESCRIPTION
**Pull request by jiaqiluo**
Issue: https://github.com/rancher/charts-build-scripts/issues/45

The following cases have been tested:
the package.yaml contains
- only version
- only packageVersion
- both version and packageVersion
- none of them

### CODE REVIEW

File: pkg/options/package.go [hunk 1]
```diff
@@ -13,8 +13,10 @@ import (
 // PackageOptions represent the options presented to users to be able to configure the way a package is built using these scripts
 // The YAML that corresponds to these options are stored within packages/<package-name>/package.yaml for each package
 type PackageOptions struct {
+	// Version represents the version of the package. It will override other values if it exists
+	Version string `yaml:"version,omitempty"`
 	// PackageVersion represents the current version of the package. It needs to be incremented whenever there are changes
-	PackageVersion int `yaml:"packageVersion" default:"0"`
+	PackageVersion int `yaml:"packageVersion,omitempty"`
```
Comment (aiyengar2) [inline pkg/options/package.go:19]: Why is this change necessary? Wouldn't the default behavior be to set the packageVersion to 0 regardless?

File: pkg/charts/package.go [hunk 2]
```diff
@@ -17,8 +17,10 @@ type Package struct {
 
 	// Name is the name of the package
 	Name string `yaml:"name"`
+	// Version represents the version of the package. It will override other values if it exists
+	Version string `yaml:"version,omitempty"`
 	// PackageVersion represents the current version of the package. It needs to be incremented whenever there are changes
-	PackageVersion int `yaml:"packageVersion"`
+	PackageVersion int `yaml:"packageVersion,omitempty"`
```
Comment (aiyengar2) [inline pkg/charts/package.go:23]: I think it would make more sense to do the parsing for Semver on trying to parse the PackageOptions. What do you think?

e.g. Let's import `"github.com/blang/semver"` into this file and use `*semver.Version` for the Version instead and change all the signatures for Version to take in a `*semver.Version` as well. A pointer is better since we can check if it is nil.

File: pkg/charts/parse.go [hunk 3]
```diff
@@ -96,6 +100,7 @@ func GetPackage(rootFs billy.Filesystem, name string) (*Package, error) {
 		Chart: chart,
 
 		Name:             name,
+		Version:          packageOpt.Version,
```
Comment (aiyengar2) [inline pkg/charts/parse.go:103]: Use `semver.Make(packageOpt.version)`

File: pkg/charts/parse.go [hunk 4]
```diff
@@ -79,6 +79,10 @@ func GetPackage(rootFs billy.Filesystem, name string) (*Package, error) {
 	if err != nil {
 		return nil, err
 	}
+	// version and packageVersion can not exist at the same time although both are optional
+	if packageOpt.Version != "" && packageOpt.PackageVersion != 0 {
+		return nil, fmt.Errorf("Cannot have both version and packageVersion at the same time")
+	}
```
Comment (aiyengar2) [inline pkg/charts/parse.go:85]: The fact that the `PackageVersion` can no longer be specified as 0 seems like a breaking change to users since this was previously allowed.

Instead of having the `packageOpt.PackageVersion` be an int, can you change it to a pointer to an int? That way, the omitempty logic you have below would also work.

Comment (aiyengar2) [reply pkg/charts/parse.go:85]: Wd should also change `version` to be a pointer, since it is also an optional field that can be nil. That way, this statement just checks whether both are non-nil.

File: pkg/helm/export.go [hunk 5]
```diff
@@ -41,13 +41,24 @@ func ExportHelmChart(rootFs, fs billy.Filesystem, helmChartPath string, packageV
...
+			return fmt.Errorf("Maximum number of packageVersions is %d, found %d", MaxPatchNum, packageVersion)
+		}
+		chartVersionSemver.Patch = PatchNumMultiplier*chartVersionSemver.Patch + uint64(packageVersion)
+	} else if version != "" {
+		chartVersionSemver, err = semver.Parse(version)
+		if err != nil {
+			return fmt.Errorf("Cannot parse version %s as an valid semver: %s", version, err)
+		}
+	} else {
+		logrus.Infof("No version or packageVersion is provided in package.yaml, using the chart version with patch files applied")
```
Comment (aiyengar2) [inline pkg/helm/export.go:56]: I think we can drop this log

File: pkg/helm/export.go [hunk 6]
```diff
@@ -41,13 +41,24 @@ func ExportHelmChart(rootFs, fs billy.Filesystem, helmChartPath string, packageV
...
+		// Add packageVersion as string, preventing errors due to leading 0s
+		if uint64(packageVersion) >= MaxPatchNum {
+			return fmt.Errorf("Maximum number of packageVersions is %d, found %d", MaxPatchNum, packageVersion)
+		}
+		chartVersionSemver.Patch = PatchNumMultiplier*chartVersionSemver.Patch + uint64(packageVersion)
+	} else if version != "" {
+		chartVersionSemver, err = semver.Parse(version)
+		if err != nil {
+			return fmt.Errorf("Cannot parse version %s as an valid semver: %s", version, err)
+		}
```
Comment (aiyengar2) [inline pkg/helm/export.go:54]: We shouldn't even try parsing `semver.Make(chart.Metadata.Version)` if the version is specified. Can you move that call into the packageVersion check?

We should probably restructure this as:

1. If version, take it.
2. Else, parse the chart.Metadata.Version. If packageVersion, append it.

File: pkg/helm/export.go [hunk 7]
```diff
@@ -41,13 +41,24 @@ func ExportHelmChart(rootFs, fs billy.Filesystem, helmChartPath string, packageV
...
+	} else {
+		logrus.Infof("No version or packageVersion is provided in package.yaml, using the chart version with patch files applied")
 	}
-	chartVersionSemver.Patch = PatchNumMultiplier*chartVersionSemver.Patch + uint64(packageVersion)
-	if len(upstreamChartVersion) > 0 {
+
+	if len(upstreamChartVersion) > 0 && upstreamChartVersion != chartVersionSemver.String() {
 		// Add buildMetadataFlag for forked charts
+		logrus.Infof("The chart's version is different from the upstream chart's, appending the upstream's version to its meatadata")
```
Comment (aiyengar2) [inline pkg/helm/export.go:61]: Don't think this log statement is necessary

**Review by aiyengar2 — CHANGES_REQUESTED** [review]

File: pkg/options/package.go (hunk 1 above)
Comment (jiaqiluo) [reply pkg/options/package.go]: The change is discarded.

**Review by aiyengar2 — APPROVED** [review]
LGTM

File: pkg/charts/parse.go [hunk 8]
```diff
@@ -79,6 +80,19 @@ func GetPackage(rootFs billy.Filesystem, name string) (*Package, error) {
 	if err != nil {
 		return nil, err
 	}
+	// version and packageVersion can not exist at the same time although both are optional
+	if packageOpt.Version != nil && packageOpt.PackageVersion != nil {
+		return nil, fmt.Errorf("Cannot have both version and packageVersion at the same time")
+	}
+	var version *semver.Version
+	if packageOpt.Version != nil {
+		temp, err := semver.Make(*packageOpt.Version)
```
Comment (cbron) [inline pkg/charts/parse.go:89]: Might run this through Matt's validator as well: https://github.com/mattfarina/semver-isvalid/blob/main/pkg/semver/valid.go

**Review by cbron — APPROVED** [review]
lgtm, comment is up for discussion.

File: pkg/charts/parse.go (hunk 8 above)
Comment (jiaqiluo) [reply pkg/charts/parse.go:89]: `semver.Make` function also does the validation as a part of the making process. 
I skim through both implementations and do not see many differences in the items/rules they check.  

We can use Matt's validator for sure if we want.
